src/modules/csm/checkpoint.py

```diff
@@ -0,0 +1,246 @@
+                checkpoint_epochs = []
+        logger.info({"msg": f"Checkpoints to process: {len(self.checkpoints)}"})
+        return self
+
+    def __next__(self):
+        if not self.checkpoints:
+            raise StopIteration
+        return self.checkpoints.pop(0)
+
+    def _get_adjusted_r_epoch(self, r_epoch: EpochNumber, finalized_epoch: EpochNumber):
+        processing_delay = finalized_epoch - self.l_epoch
+        if processing_delay < self.MIN_CHECKPOINT_STEP and finalized_epoch < r_epoch:
+            logger.info(
+                {
+                    "msg": f"Minimum checkpoint step is not reached, current delay is {processing_delay} epochs",
+                    "finalized_epoch": finalized_epoch,
+                    "l_epoch": self.l_epoch,
+                    "r_epoch": r_epoch,
+                }
+            )
+            raise ValueError('Minimum checkpoint step is not reached yet')
+        adjusted_r_epoch = min(r_epoch, EpochNumber(finalized_epoch - 1))
+        if r_epoch != adjusted_r_epoch:
+            logger.warning(
+                {
+                    "msg": f"Right border epoch of checkpoints iterator is recalculated according to the finalized epoch. "
+                    f"Before: {r_epoch} After: {adjusted_r_epoch}"
+                }
+            )
+        return adjusted_r_epoch
+
+
+class CheckpointProcessor:
+    cc: ConsensusClient
+    converter: Web3Converter
+
+    state: State
+    finalized_blockstamp: BlockStamp
+
+    def __init__(self, cc: ConsensusClient, state: State, converter: Web3Converter, finalized_blockstamp: BlockStamp):
+        self.cc = cc
+        self.converter = converter
+        self.state = state
+        self.finalized_blockstamp = finalized_blockstamp
+
+    def exec(self, checkpoint: Checkpoint) -> int:
+        logger.info(
+            {"msg": f"Processing checkpoint for slot {checkpoint.slot} with {len(checkpoint.duty_epochs)} epochs"}
+        )
+        unprocessed_epochs = [e for e in checkpoint.duty_epochs if e in self.state.unprocessed_epochs]
+        if not unprocessed_epochs:
+            logger.info({"msg": "Nothing to process in the checkpoint"})
+            return 0
+        block_roots = self._get_block_roots(checkpoint.slot)
+        duty_epochs_roots = {
+            duty_epoch: self._select_block_roots(duty_epoch, block_roots, checkpoint.slot)
+            for duty_epoch in unprocessed_epochs
+        }
+        self._process(unprocessed_epochs, duty_epochs_roots)
+        return len(unprocessed_epochs)
+
+    def _get_block_roots(self, checkpoint_slot: SlotNumber):
+        logger.info({"msg": f"Get block roots for slot {checkpoint_slot}"})
+        # checkpoint for us like a time point, that's why we use slot, not root
+        br = self.cc.get_state_block_roots(checkpoint_slot)
+        # replace duplicated roots to None to mark missed slots
+        return [None if br[i] == br[i - 1] else br[i] for i in range(len(br))]
```

Comment:
Remove this filtering

Reply:
we can't do that. we need list without duplicates already here because we can't filter them in `_select_block_roots`